beetsplug/duplicates.py

```diff
@@ -197,7 +200,20 @@ def _checksum(self, item, prog):
         output as flexattr on a key that is the name of the program, and
         return the key, checksum tuple.
         """
-        args = [p.format(file=item.path) for p in shlex.split(prog)]
+        # On Python 3, we need to construct the command to invoke as a
+        # Unicode string. On Unix, this is a little unfortunate---the OS is
+        # expecting bytes---so we use surrogate escaping and decode with the
+        # argument encoding, which is the same encoding that will then be
+        # *reversed* to recover the same bytes before invoking the OS. On
+        # Windows, we want to preserve the Unicode filename "as is."
+        if not six.PY2:
+            if platform.system() == 'Windows':
+                args = [p.format(file=item.path.decode(util._fsencoding()))
+                        for p in shlex.split(prog)]
+            else:
+                args = [p.format(file=item.path.decode(util.arg_encoding(),
+                        'surrogateescape')) for p in shlex.split(prog)]
+
```

Comment:
Apparently I failed to save the comment. The following is what I had written:

---

I think this would be a great time to abstract these conversions into a function in `beets.util`, along the lines of
```python
# beets/util/__init__.py

"""Substitute a file path in a commandline argument for external tools.

On Python 3, we need to construct the command to invoke as a
Unicode string. On Unix, this is a little unfortunate---the OS is
expecting bytes---so we use surrogate escaping and decode with the
argument encoding, which is the same encoding that will then be
*reversed* to recover the same bytes before invoking the OS. On
Windows, we want to preserve the Unicode filename "as is."
"""
def format_commandline_path(fmt, path):
    if six.PY2:
        # On Python 2, substitute the bytestring path into a bytestring
        # template.
        return fmt.format(file=path)
    else:
        # On Python 3, the template is a Unicode string, which only supports
        # substitution of Unicode variables.
        if platform.system() == 'Windows':
            return fmt.format(file=item.path.decode(util._fsencoding()))
        else:
            return fmt.format(file=item.path.decode(util.arg_encoding(),
                              'surrogateescape'))
```


```python
# convert.py, duplicates.py
from beets import util
args = [util.format_commandline_argument(p, file=item.path) for p in shlex.split(prog)]
```
Note that your code misses the Python 2 code path, which would lead to a crash (and in principle, we had the intention to support Python 2 until the next release).

---

Actually, this doesn't work for code sharing with the convert plugin, because the latter uses a different templating mechanism... The following could work (I didn't test it, though):
```python
# beets/util/__init__.py

"""Prepare a path for substitution into commandline template.

On Python 3, we need to construct the subprocess commands to invoke as a
Unicode string. On Unix, this is a little unfortunate---the OS is
expecting bytes---so we use surrogate escaping and decode with the
argument encoding, which is the same encoding that will then be
*reversed* to recover the same bytes before invoking the OS. On
Windows, we want to preserve the Unicode filename "as is."
"""
def decode_commandline_path(path):
    if six.PY2:
        # On Python 2, substitute the bytestring directly into the template.
        return path
    else:
        # On Python 3, the template is a Unicode string, which only supports
        # substitution of Unicode variables.
        if platform.system() == 'Windows':
            return path.decode(util._fsencoding())
        else:
            return path.decode(util.arg_encoding(), 'surrogateescape')
```


```python
# duplicates.py
args = [p.format(file=util.decode_commandline_path(item.path)) for p in shlex.split(prog)]
```

```python
# convert.py
source = util.decode_commandline_path(source)
dest = util.decode_commandline_path(dest)
```